### PR TITLE
Changes to XCode task to publish test results

### PR DESCRIPTION
**Pull request by prawalagarwal**
Description: Currently in the xcode task, we provide the option to run the xcode project tests and generate junit output using xctool, but this test result file is not published to TFS. If the user is specifying the output result file, we should publish it without using the Publish Results Task.

Solution: A field has been added to the task which asks the users if he/she wants to publish the test results. The generated file is published if the checkbox is checked.

Testing: Manual

### CODE REVIEW

File: Tasks/Xcode/task.json [hunk 1]
```diff
@@ -213,8 +213,17 @@
             "required":false,
             "helpMarkDown": "Test reporter format to use when \"test\" action is specified and \"Use xctool\" is checked. Specify \"junit:output-file-path-here.xml\" to generate a file format compatible with the Publish Test Results task. When specified, \"plain\" is automatically added as well. Requires xctool be installed on agent hosts. See [xctool](https://github.com/facebook/xctool) for details.",
             "groupName":"advanced"
+        },
+		{
+            "name":"publishJUnitResults",
+            "type":"boolean",
+            "label":"Publish to VSO/TFS",
+            "required":false,
+            "defaultValue":true,
```
Comment (allendm-msft) [inline Tasks/Xcode/task.json]: this should match the default for xctool otherwise we will give a warning by default

File: Tasks/Xcode/task.json [hunk 2]
```diff
@@ -213,8 +213,17 @@
             "required":false,
             "helpMarkDown": "Test reporter format to use when \"test\" action is specified and \"Use xctool\" is checked. Specify \"junit:output-file-path-here.xml\" to generate a file format compatible with the Publish Test Results task. When specified, \"plain\" is automatically added as well. Requires xctool be installed on agent hosts. See [xctool](https://github.com/facebook/xctool) for details.",
             "groupName":"advanced"
+        },
+		{
+            "name":"publishJUnitResults",
+            "type":"boolean",
+            "label":"Publish to VSO/TFS",
```
Comment (allendm-msft) [inline Tasks/Xcode/task.json]: You can no longer use VSO

File: Tasks/Xcode/xcodebuild2.js [hunk 3]
```diff
@@ -221,3 +229,27 @@ function removeExecOutputNoise(input) {
...
+      var buildFolder = tl.getVariable('agent.buildDirectory');
+      var allFiles = tl.find(buildFolder);
+      var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
+    }
+    else {
+      tl.debug('No pattern found in testResultsFiles parameter');
+      var matchingTestResultsFiles = [testResultsFiles];
+    }
+
+    if(!matchingTestResultsFiles) {
```
Comment (allendm-msft) [inline Tasks/Xcode/xcodebuild2.js]: if publish is checked but xctool is not checked give the right warning with the resolution ie enable xctool for publish to work

File: Tasks/Xcode/xcodebuild2.js [hunk 4]
```diff
@@ -221,3 +229,27 @@ function removeExecOutputNoise(input) {
...
+      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
```
Comment (allendm-msft) [inline Tasks/Xcode/xcodebuild2.js]: check if there is localization support in js

File: Tasks/Xcode/xcodebuild2.js [hunk 5]
```diff
@@ -83,6 +87,10 @@ function processInputs() {
 	xcb.arg('-configuration');
 	xcb.arg(tl.getInput('configuration', true));
 	
+	//Test Results publish inputs
+	publishResults = (tl.getInput('publishJUnitResults', true) == "true");
+    testResultsFiles = tl.getInput('xctoolReporter', false).split(":")[1].trim();
```
Comment (allendm-msft) [inline Tasks/Xcode/xcodebuild2.js]: fix the indentation

File: Tasks/Xcode/xcodebuild2.js [hunk 6]
```diff
@@ -83,6 +87,10 @@ function processInputs() {
 	xcb.arg('-configuration');
 	xcb.arg(tl.getInput('configuration', true));
 	
+	//Test Results publish inputs
+	publishResults = (tl.getInput('publishJUnitResults', true) == "true");
```
Comment (allendm-msft) [inline Tasks/Xcode/xcodebuild2.js]: in js its better to be explicit with the type and use ===

File: Tasks/Xcode/task.json (hunk 2 above)
Comment (bryanmacfarlane) [inline Tasks/Xcode/task.json]: change to just 'VSTS'.  We will be renaming but we shouldn't be adding new VSO references

File: Tasks/Xcode/task.json (hunk 1 above)
Comment (bryanmacfarlane) [inline Tasks/Xcode/task.json]: This should default to false.

File: Tasks/Xcode/task.json [hunk 7]
```diff
@@ -213,8 +213,17 @@
...
+            "required":false,
+            "defaultValue":true,
+            "groupName":"advanced",
+            "helpMarkDown":"Select this option to publish JUnit Test results produced above using xctool to VSO/TFS."
```
Comment (bryanmacfarlane) [inline Tasks/Xcode/task.json]: Same 'VSTS'

File: Tasks/Xcode/xcodebuild2.js [hunk 8]
```diff
@@ -26,7 +26,11 @@ processInputs() 													// Process inputs to task and create xcv, xcb, impo
 	.then(function(code) {
 		return xcv.exec();											// Print version of xcodebuild / xctool
 	})			
-	.then(execBuild)												// Run main xcodebuild / xctool task
+	.then(execBuild)                                                // Run main xcodebuild / xctool task
+	.then(function(code) {                                          // publish test results
+		publishTestResults(publishResults, testResultsFiles);
+		return code;
```
Comment (bryanmacfarlane) [inline Tasks/Xcode/xcodebuild2.js]: the variable code is the return code from ExecBuild here, correct?

File: Tasks/Xcode/xcodebuild2.js (hunk 6 above)
Comment (bryanmacfarlane) [inline Tasks/Xcode/xcodebuild2.js]: The new task lib which is linked into the task has tl.getBoolInput method which does what you're doing here.

File: Tasks/Xcode/xcodebuild2.js (hunk 5 above)
Comment (bryanmacfarlane) [inline Tasks/Xcode/xcodebuild2.js]: This will null ref on .split if xctoolReporter input is not supplied (which I imagine would be the default)

Comment (bryanmacfarlane) [inline Tasks/Xcode/xcodebuild2.js]: You should probably guard the [1] as well.  That will null ref on the trim() call
```
'foo'.split(':')[1].trim()
```

File: Tasks/Xcode/xcodebuild2.js (hunk 8 above)
Comment (scottdallamura) [inline Tasks/Xcode/xcodebuild2.js]: Should we check the code and bail out if it indicates a failure?

Comment (prawalagarwal) [inline Tasks/Xcode/xcodebuild2.js]: Bryan: Yes its the return code from ExecBuild
Scott: We usually don't bail out in java tasks when publishing fails (we try to print appropriate error messages). Followed the same here. Also will the .fail at the end of calls not handle it?

File: Tasks/Xcode/xcodebuild2.js (hunk 6 above)
Comment (prawalagarwal) [inline Tasks/Xcode/xcodebuild2.js]: Using the method and replacing all other similar calls.

File: Tasks/Xcode/xcodebuild2.js (hunk 4 above)
Comment (prawalagarwal) [inline Tasks/Xcode/xcodebuild2.js]: Currently there is no localization support.

**Pull request merged**